### PR TITLE
New pipeline

### PR DESCRIPTION
**Pull request by rvodden**
New & exciting docker based workflow. As yet mostly untested and undocumented (that's tomorrows job)

## please don't merge this yet!

Things to do:

- [x] Get build passing in travis
- [x] Get build passing in appveyor
- [x] Re-establish matrix build in travis
- [x] Remove dirty 60 second pause in docker flow (find way to detect that apache2 is up)
- [x] Integrate static code inspection
- [x] Write documentation

Things to do post merge:
- [ ] Grab artefacts (logs etc.) from containers and put them somewhere (cat to stdout on failure?)
- [ ] Publish test results to appveyor & travis (and the SCI engines?)
- [ ] Integrate the documentation publication into phing

Things to consider:
- [ ] Coverage
- [ ] Videos/Screenshots
- [ ] Use windows containers in appveyor

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,68 +2,46 @@ sudo: false
...
+env:
+  - WORDHAT_DRIVER=wpphp
+  - WORDHAT_DRIVER=wpcli
+  
+services:
+  - docker
+
+#before_cache:
+#  # Save tagged docker images
+#  - mkdir -p $HOME/docker && docker images -a --filter='dangling=false' --format '{{.Repository}}:{{.Tag}} {{.ID}}' | xargs -n 2 -t sh -c 'test -e $HOME/docker/$1.tar.gz || docker save $0 | gzip -2 > $HOME/docker/$1.tar.gz'
```
Comment (paulgibbs) [inline .travis.yml]: Is this `$HOME/docker` stuff going to be used? Does it work now?

Comment (rvodden) [reply .travis.yml]: Those commented out things are for a future optimization at the moment travis is pulling in all the images from scratch each time, which is very safe, but not very quick. That $HOME/docker will be a place to put and restore images from travis cache.

File: build/templates/Dockerfile.template [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+RUN sed -i '1 s/\*\:80/\*\:8080/' /etc/apache2/sites-available/000-default.conf
+RUN sed -i 's/Listen 80/Listen 8080/' /etc/apache2/ports.conf
+EXPOSE 8080
+
+RUN apt-get -y update && apt-get -y --force-yes install gnupg && apt-key update  
+RUN apt-get install -y --force-yes netcat mysql-client zip unzip
+
+RUN curl https://raw.githubusercontent.com/composer/getcomposer.org/master/web/installer | php -- && mv composer.phar /usr/local/bin/composer
+#RUN yes | pecl install xdebug \
+#    && echo "zend_extension=$(find /usr/local/lib/php/extensions/ -name xdebug.so)" > /usr/local/etc/php/conf.d/xdebug.ini \
```
Comment (paulgibbs) [inline build/templates/Dockerfile.template]: Do we need xdebug?

Comment (rvodden) [reply build/templates/Dockerfile.template]: its not been release for 7.2 yet, and it breaks phpstan, so I've left it out for the moment.

File: composer.json [hunk 3]
```diff
@@ -1,49 +1,58 @@
...
-    "phpcs":   "phpcs --standard=phpcs-ruleset.xml -p -s -v -n src --extensions=php",
-    "phpcbf":  "phpcbf --standard=phpcs-ruleset.xml -n src",
-    "phpstan": "vendor/bin/phpstan analyse -c phpstan.neon --level=1 src",
-    "commit": [
-      "@phpcs",
-      "@phpstan"
-    ]
-  }
-}
+	"name" : "paulgibbs/behat-wordpress-extension",
```
[the remark this answers is not in the document]
Comment (rvodden) [reply composer.json]: that's odd - I though my ide was picking up .editorconfig. I'll fix both :)

File: build/templates/Dockerfile.template (hunk 2 above)
Comment (paulgibbs) [reply build/templates/Dockerfile.template]: ok nice

File: .travis.yml (hunk 1 above)
Comment (paulgibbs) [reply .travis.yml]: nice plan!

**Comment by paulgibbs** [comment]
LGTM!

**Pull request merged**